File: java/src/leetcode/p003/Solution.java
package leetcode.p003;

import java.util.HashMap;

class Solution {

    public static void main(String [] args){
        Solution sol = new Solution();
        System.out.println(sol.lengthOfLongestSubstring("abcabcbb"));
    }

    public int lengthOfLongestSubstring(String s) {
        if (s == null) return 0;
        if (s.length() == 0) return 0;
        HashMap<Character, Integer> hmap = new HashMap<>();
        int maxLength = 0;
        int curLength = 0;
        int cur_cutoff = 0;
        for (int i = 0; i < s.length(); i++){
            Character cur = s.charAt(i);
            if (hmap.containsKey(cur)){
                Integer posOld = hmap.get(cur);
                if (posOld < cur_cutoff)
                {
                    curLength++;
                    if (curLength > maxLength)
                        maxLength = curLength;
                    hmap.put(cur,i);
                } else {
                    curLength = i - posOld;
                    hmap.put(cur, i);
                    cur_cutoff = posOld;
                    if (curLength > maxLength)
                        maxLength = curLength;
                }
            } else {
                curLength++;
                if (curLength > maxLength)
                    maxLength = curLength;
                hmap.put(cur,i);
            }
        }

        return maxLength;
    }
}

File: scala/src/leetcode/p003/Solution.scala
package leetcode.p003

object Solution {

  @main
  def main() = {
    println(lengthOfLongestSubstring("baab!bb"))
  }

  def lengthOfLongestSubstring(s: String): Int = {

    var current = scala.collection.mutable.Set.empty[Char]

    var max = 0

    var i = 0
    var j = 0

    while (i < s.length()) {
      val ch = s.charAt(i)

      if (!current.contains(ch)) {
        current.add(ch)
        if (current.size > max)
            max = current.size 
      } else {
        while (s.charAt(j) != ch){
            val prev = s.charAt(j)
            current.remove(prev)
            j = j+1
        }
        j = j + 1
      }
      i = i + 1
    }


    max
  }
}
